### PR TITLE
Allow log server to return consistency proofs where first=second

### PR DESCRIPTION
**Pull request by Martin2112**
This is for additional compatiblity with the older C++ log server code. Needs a corresponding change to CTFE, which will follow:

https://github.com/google/certificate-transparency-go/pull/74

### CODE REVIEW

File: server/log_rpc_server_test.go [hunk 1]
```diff
@@ -1231,38 +1234,91 @@ func TestGetConsistencyProofCommitFails(t *testing.T) {
...
-		t.Fatalf("expected proof: %v but got: %v", expectedProof, response.Proof)
+	server := NewTrillianLogRPCServer(registry, fakeTimeSource)
+
+	_, err := server.GetConsistencyProof(context.Background(), &getConsistencyProofRequest54)
+	if err == nil {
+		t.Fatal("incorrectly returned a proof for out of range tree size")
+	}
+}
+
+func TestGetConsistencyProof(t *testing.T) {
```
Comment (AlCutter) [inline server/log_rpc_server_test.go]: Seems like a lot of these tests could be collapsed into a table-driven one which includes a `wantError` flag, no?
Perhaps that could be done in a follow-up, though.

File: server/validate.go [hunk 2]
```diff
@@ -51,7 +51,7 @@ func validateGetConsistencyProofRequest(req *trillian.GetConsistencyProofRequest
 	if req.SecondTreeSize <= 0 {
 		return status.Errorf(codes.InvalidArgument, "GetConsistencyProofRequest.SecondTreeSize: %v, want > 0", req.SecondTreeSize)
 	}
-	if req.SecondTreeSize <= req.FirstTreeSize {
+	if req.SecondTreeSize < req.FirstTreeSize {
 		return status.Errorf(codes.InvalidArgument, "GetConsistencyProofRequest.FirstTreeSize: %v < GetConsistencyProofRequest.SecondTreeSize: %v, want > ", req.FirstTreeSize, req.SecondTreeSize)
```
Comment (AlCutter) [inline server/validate.go]: want >=

File: server/log_rpc_server_test.go (hunk 1 above)
Comment (Martin2112) [reply server/log_rpc_server_test.go]: Yes it could be done but seemed out of scope for this change.

File: server/validate.go (hunk 2 above)
Comment (Martin2112) [reply server/validate.go]: Done.

**Comment by Martin2112** [comment]
Finally got Travis to run. PTAL.

**Comment by Martin2112** [comment]
@daviddrysdale can you take a quick look? I've got a followup in progress to address the test comment Al made.

File: server/log_rpc_server_test.go [hunk 3]
```diff
@@ -1231,38 +1234,91 @@ func TestGetConsistencyProofCommitFails(t *testing.T) {
...
+
+		registry := extension.Registry{
+			AdminStorage: mockAdminStorage(ctrl, testCases[i].LogId),
+			LogStorage:   mockStorage,
+		}
+		server := NewTrillianLogRPCServer(registry, fakeTimeSource)
+
+		response, err := server.GetConsistencyProof(context.Background(), &testCases[i])
+		if err != nil {
+			t.Fatalf("failed to get consistency proof: %v", err)
```
Comment (daviddrysdale) [inline server/log_rpc_server_test.go]: `t.Errorf` and `continue` so the loop continues
(OK to defer if you've got a follow-up PR that makes things more table-driven overall)

File: server/log_rpc_server_test.go [hunk 4]
```diff
@@ -1231,38 +1234,91 @@ func TestGetConsistencyProofCommitFails(t *testing.T) {
...
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	// getConsistencyProofRequest7 - tests a normal request which should succeed
+	// getConsistencyProofRequest44 - tests an edge condition we used to reject but now support
+	// for compatibility with the older C++ log servers.
+	testCases := []trillian.GetConsistencyProofRequest{getConsistencyProofRequest7, getConsistencyProofRequest44}
+	nodeIDs := [][]storage.NodeID{nodeIdsConsistencySize4ToSize7, {}}
+	hashes := [][][]byte{{[]byte("nodehash")}, {}}
+	nodes := [][]storage.Node{{{NodeID: stestonly.MustCreateNodeIDForTreeCoords(2, 1, 64), NodeRevision: 3, Hash: []byte("nodehash")}}, {}}
```
Comment (daviddrysdale) [inline server/log_rpc_server_test.go]: Better to put all the per-test case info into a single struct:
```go
vars tests = []struct {
  req trillian.GetConsistencyProofRequest
  nodeIDs []storage.NodeID
  hashes [][]byte
  nodes []storage.Node
}{
  {
    req: &getConsistencyProofRequest7,
    nodeIDs: nodeIdsConsistencySize4ToSize7,
    hashes: [][]byte{{[]byte("nodehash")},
    nodes: []storage.Node{{NodeID: stestonly.MustCreateNodeIDForTreeCoords(2, 1, 64), NodeRevision: 3, Hash: []byte("nodehash")}},
  },
  {
    req: &getConsistencyProofRequest44,
  },
}
for _, test := range tests {
   ...
```

(drop if this is coming in follow-up PR)

Comment (Martin2112) [reply server/log_rpc_server_test.go]: Yeah I've got this done but the branch is based off this one.

File: server/log_rpc_server_test.go (hunk 3 above)
Comment (Martin2112) [reply server/log_rpc_server_test.go]: Yes, upcoming PR moves the consistency proof tests into one table test.
